src/adapters/webpush/mod.rs

```diff
@@ -0,0 +1,359 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+extern crate serde_json;
+
+mod crypto;
+mod db;
+
+use controller::Controller;
+use foxbox_adapters::adapter::*;
+use foxbox_taxonomy::api::{ Error, InternalError };
+use foxbox_taxonomy::values::{ Range, Type, Value, Json };
+use foxbox_taxonomy::services::*;
+use foxbox_taxonomy::util::Id;
+use hyper::header::{ ContentEncoding, Encoding };
+use hyper::Client;
+use hyper::client::Body;
+use rusqlite::{ self };
+use std::collections::{ HashMap, HashSet };
+use std::sync::Arc;
+use std::thread;
+use transformable_channels::mpsc::*;
+
+header! { (Encryption, "Encryption") => [String] }
+header! { (EncryptionKey, "Encryption-Key") => [String] }
+
+static ADAPTER_NAME: &'static str = "WebPush adapter (built-in)";
+static ADAPTER_VENDOR: &'static str = "team@link.mozilla.org";
+static ADAPTER_VERSION: [u32;4] = [0, 0, 0, 0];
```

Comment:
For the moment, version numbers are entirely unused and entirely up to renegotiation.